core/src/main/scala/com/github/johnynek/paiges/Doc.scala

```diff
@@ -287,10 +289,32 @@ object Doc {
       case Nil => Empty
       case x :: Nil => x
       case x :: y :: tail =>
+        /**
+         * The cost of this algorithm c(n) for list of size n.
+         * note that c(n) = 2 * c(n-1) + k
+         * for some constant.
+         * so, c(n) - c(n-1) = c(n-1) + k
+         * which means that
+         * c(n) = (0 until n).map(c(_)).sum + nk
+         *
+         * which is exponential in n (O(2^n))
+         *
+         * making the second parameter in the union lazy would fix this.
+         * that seems an expensive fix for a single combinator. Maybe
+         * there is an alternative way to express this that is not
+         * exponential.
+         *
+         * On top of this difficulty, this formulation creates
+         * Union nodes that violate the invariant that Union(a, b)
+         * means a == flatten(b). It still has flatten(a) == flatten(b),
+         * however. This fact seems to complicate comparison of Doc
+         * which is valuable.
+         */
```

Comment:
I agree that it would be nice to come up with a more efficient alternate formulation.